### DISCUSSION TITLE
[SHOW & TELL] Printing labels automatically through react-native app

### DISCUSSION THREAD

**Discussion by Zontex** [inventree/InvenTree#2715] [Show and tell]
After implementing smart LEDs solution, it's time to hack the thermal label printer.

The thermal label printer doesn't have any Linux driver so I had to come up with python code to control it, can print pdf labels (one or many) currently, it's working through the micro flask API I've made but I will integrate it into the plugin later on.

Python code originally by @abrasive but I modified it to support multiple labels at once: https://github.com/OmnIoT/pdf2tspl
label printer connected directly to Raspberry Pi (OmniKit) that hosts Inventree locally.

Pictures from the mobile app:

![Screen Shot 2022-03-05 at 9 11 23 PM](https://user-images.githubusercontent.com/19342533/156884519-b45a2431-b005-4ab4-8a10-aee0893fd5c7.png)
![Screen Shot 2022-03-05 at 9 11 45 PM](https://user-images.githubusercontent.com/19342533/156884537-19ed93f2-600a-4557-9958-d6668de81457.png)

Thermal label printer (N41, cheap but working amazingly well):

![11381646486037_ pic](https://user-images.githubusercontent.com/19342533/156884622-cd1af9dd-5865-440f-88bb-91b337e112dc.jpg)

.

**Comment by Zontex** [comment]
@matmair @SchrodingersGat would love to hear your feedback guys on this :)

**Reply by matmair** [reply]
@zontex looks interesting. Any plans to release / source the app? I have been looking into react-native the last while and would be interested on how you implemented it.

**Reply by Zontex** [reply]
> @Zontex looks interesting. Any plans to release/source the app? I have been looking into react-native the last while and would be interested in how you implemented it.

yes absolutely but it will take me a while (perhaps another 2-3 weeks) currently a lot of things hard-coded such the server and they don't pass correctly through navigation, the point was to create "quick and dirty POC", I'm going to focus on making it a more consumer usable app this week and perhaps will release it later on.

**Comment by SchrodingersGat** [comment]
@Zontex you have inspired me!

InvenTree now directly supports plugins for printing labels: https://github.com/inventree/InvenTree/pull/2768

This means that if you write a plugin for your printer, you can print labels to that printer directly from the web interface:

![image](https://user-images.githubusercontent.com/10080325/160218778-1de1eb5d-44dc-4259-9831-ee7621b843a2.png)

Also, in the next app release, you can print labels directly from the app.

Would you be willing to look at supporting your printer driver library as an inventree plugin?

I've hacked together a plugin for the Brother QL/PT series of printers - [source code here](https://github.com/inventree/inventree-brother-plugin) - the implementation is pretty straight forward!

Docs here: https://inventree.readthedocs.io/en/latest/extend/plugins/label/

**Reply by Zontex** [reply]
That looks amazing, I will absolutely try to integrate my plugin, let me go over the weekend and upload some code to see if it can work. is this version in the master branch already?

**Reply by SchrodingersGat** [reply]
Yep, master branch now supports label printing via plugins!

**Reply by matmair** [reply]
@SchrodingersGat very nice!
